src/platforms/ctxlink/WiFi_Server.c

```diff
@@ -0,0 +1,1511 @@
+		swo_trace_addr.sin_addr.s_addr = 0;
+		swo_trace_addr.sin_family = AF_INET;
+		swo_trace_addr.sin_port = _htons(SWO_TRACE_SERVER_PORT);
+		int8_t result = bind(swo_trace_server_socket, (struct sockaddr *)&swo_trace_addr, sizeof(swo_trace_addr));
+		if (result != SOCK_ERR_NO_ERROR) {
+			return;
+		}
+		swo_trace_tcp_serverstate = sm_listening;
+		break;
+	}
+	case sm_listening: {
+		//
+		// No need to perform any flush.
+		// TCP data in TX FIFO will automatically transmit itself after it accumulates for a while.
+		// If you want to decrease latency (at the expense of wasting network bandwidth on TCP overhead),
+		// perform an explicit flush via the TCPFlush() API.
+		break;
+	}
+	case sm_closing: {
+		// Close the socket connection.
+		close(swo_trace_server_socket);
+		swo_trace_server_socket = SOCK_ERR_INVALID;
+		swo_trace_tcp_serverstate = sm_idle;
+		break;
+	}
+	}
+}
+
+void wifi_setup_swo_trace_server(void)
+{
+	//
+	// Is there a UART client connected?
+	//
+	if (uart_debug_client_connected) {
+		// Close the connection to the client
+		close(uart_debug_client_socket);
+		uart_debug_client_socket = SOCK_ERR_INVALID; // Mark socket invalid
+		uart_debug_client_connected = false;         // No longer connected
+		user_configured_uart = false;
+	}
+	//
+	// If the UART server is up, close it timeDown
+	//
+	if (uart_debug_server_socket != SOCK_ERR_INVALID) {
+		close(uart_debug_server_socket);
+		uart_debug_server_socket = SOCK_ERR_INVALID;
+	}
+	//
+	// Set up the SWO Trace Server
+	//
+	swo_trace_tcp_serverstate = sm_home;
+}
+
+static void app_wifi_callback(uint8_t msg_type, void *msg)
+{
+	(void)msg;
+	switch (msg_type) {
+	case M2M_WIFI_DRIVER_INIT_EVENT: {
+		driver_init_complete = true;
+		break;
+	}
+
+	case M2M_WIFI_CONN_STATE_CHANGED_EVENT: {
+		tstrM2mWifiStateChanged *wifi_state = (tstrM2mWifiStateChanged *)msg;
+		if (wifi_state->u8CurrState == M2M_WIFI_CONNECTED) {
+			//
+			// If we are in http access mode there will be a connection event when the
+			// provisioning client connects, we need to ignore this event.
+			//
+			// The connection event is the provisioning client if:
+			//		http_active == true && waiting_access_point == true
+			//
+			// Otherwise it is a valid ctxLink-> AP connection
+			//
+			if (http_active && waiting_access_point) {
+				waiting_access_point = false; // Next event is the ctxLink to AP event
+			} else {
+				DEBUG_WARN("APP_WIFI_CB[%d]: Connected to AP\r\n", msg_type);
+				g_wifi_connected = true;
+			}
+		} else if (wifi_state->u8CurrState == M2M_WIFI_DISCONNECTED) {
+			DEBUG_WARN("APP_WIFI_CB[%d]: Disconnected from AP\r\n", msg_type);
+			g_wifi_connected = false;
+		} else {
+			DEBUG_WARN("APP_WIFI_CB[%d]: Unknown WiFi state change\r\n", msg_type);
+		}
+		break;
+	}
+
+	case M2M_WIFI_IP_ADDRESS_ASSIGNED_EVENT: {
+		t_wifiEventData *wifi_event_data = (t_wifiEventData *)msg;
+		if (wifi_event_data != NULL) {
+			ip_address_assigned = true;
+		}
+		ip_address_assigned = true;
```

Comment:
Is something different supposed to happen for the two sides of this branch? This looks like a potential bug.

Reply:
Fixed